File: internal/proxy/resources.go
// https://hypercall.xyz

package proxy

import (
	"context"
	"encoding/json"
	"fmt"

	"github.com/hypercall-public/mcpzip/internal/types"
)

// Resource represents an MCP resource from an upstream server.
type Resource struct {
	URI         string `json:"uri"`
	Name        string `json:"name"`
	Description string `json:"description,omitempty"`
	MimeType    string `json:"mimeType,omitempty"`
	ServerName  string `json:"server_name"`
}

// Prompt represents an MCP prompt from an upstream server.
type Prompt struct {
	Name        string          `json:"name"`
	Description string          `json:"description,omitempty"`
	Arguments   json.RawMessage `json:"arguments,omitempty"`
	ServerName  string          `json:"server_name"`
}

// PrefixURI returns a server-prefixed URI.
func PrefixURI(server, uri string) string {
	return server + types.NameSeparator + uri
}

// ParsePrefixedURI splits a prefixed URI back into server and original URI.
func ParsePrefixedURI(prefixed string) (server, uri string, err error) {
	return types.ParsePrefixedName(prefixed)
}

// ListResources returns all resources aggregated from upstream servers.
// Currently returns empty since upstream connections don't support
// resource listing yet (pending MCP SDK integration).
func (s *Server) ListResources(ctx context.Context) ([]Resource, error) {
	// Will be populated when real MCP SDK connections support resources.
	return nil, nil
}

// ReadResource reads a resource by its prefixed URI,
// routing to the correct upstream server.
func (s *Server) ReadResource(ctx context.Context, prefixedURI string) (json.RawMessage, error) {
	server, _, err := ParsePrefixedURI(prefixedURI)
	if err != nil {
		return nil, fmt.Errorf("invalid resource URI %q: %w", prefixedURI, err)
	}
	// Verify server exists.
	_ = server
	return nil, fmt.Errorf("resource reading not yet implemented (pending MCP SDK integration)")
}

// ListPrompts returns all prompts aggregated from upstream servers.
func (s *Server) ListPrompts(ctx context.Context) ([]Prompt, error) {
	return nil, nil
}

// GetPrompt gets a prompt by its prefixed name.
func (s *Server) GetPrompt(ctx context.Context, prefixedName string) (json.RawMessage, error) {
	server, _, err := types.ParsePrefixedName(prefixedName)
	if err != nil {
		return nil, fmt.Errorf("invalid prompt name %q: %w", prefixedName, err)
	}
	_ = server
	return nil, fmt.Errorf("prompt retrieval not yet implemented (pending MCP SDK integration)")
}

File: src/proxy/resources.rs
use serde::{Deserialize, Serialize};
use serde_json::Value;

use crate::error::McpzipError;
use crate::proxy::server::ProxyServer;
use crate::types;

/// An MCP resource from an upstream server.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Resource {
    pub uri: String,
    pub name: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub mime_type: Option<String>,
    pub server_name: String,
}

/// An MCP prompt from an upstream server.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Prompt {
    pub name: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub arguments: Option<Value>,
    pub server_name: String,
}

/// Returns a server-prefixed URI.
pub fn prefix_uri(server: &str, uri: &str) -> String {
    types::prefixed_name(server, uri)
}

/// Split a prefixed URI into (server, original_uri).
pub fn parse_prefixed_uri(prefixed: &str) -> Result<(&str, &str), McpzipError> {
    types::parse_prefixed_name(prefixed)
}

impl ProxyServer {
    /// List all resources (stub - returns empty).
    pub fn list_resources(&self) -> Vec<Resource> {
        Vec::new()
    }

    /// Read a resource by prefixed URI (stub - not yet implemented).
    pub fn read_resource(&self, prefixed_uri: &str) -> Result<Value, McpzipError> {
        let (_server, _uri) = parse_prefixed_uri(prefixed_uri)?;
        Err(McpzipError::Protocol(
            "resource reading not yet implemented".into(),
        ))
    }

    /// List all prompts (stub - returns empty).
    pub fn list_prompts(&self) -> Vec<Prompt> {
        Vec::new()
    }

    /// Get a prompt by prefixed name (stub - not yet implemented).
    pub fn get_prompt(&self, prefixed_name: &str) -> Result<Value, McpzipError> {
        let (_server, _name) = types::parse_prefixed_name(prefixed_name)?;
        Err(McpzipError::Protocol(
            "prompt retrieval not yet implemented".into(),
        ))
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_prefix_uri() {
        assert_eq!(
            prefix_uri("slack", "file:///channels.json"),
            "slack__file:///channels.json"
        );
    }

    #[test]
    fn test_parse_prefixed_uri() {
        let (server, uri) = parse_prefixed_uri("slack__file:///channels.json").unwrap();
        assert_eq!(server, "slack");
        assert_eq!(uri, "file:///channels.json");
    }

    #[test]
    fn test_parse_prefixed_uri_invalid() {
        assert!(parse_prefixed_uri("no-separator").is_err());
    }
}
